### ISSUE TITLE
Fix string-interpolation of lists, and normalization of string-quoting

### ISSUE THREAD

**Issue by lunelson** [sass/libsass#870]
This issue seems to have some crossover with #639 and #694. It pertains both to the normalization of string-quoting, and to the string-interpolation of lists. The test takes two lists of quoted strings, a comma-separated one and a space-separated one, each containing both double-quoted and single-quoted items, and passes them to a few interpolations, involving containing quotes and/or other characters. Corresponding spec to be PR'd forthwith.

```scss
$quoted-strings-csv: "alpha", "beta", 'gamma', 'delta';
$quoted-strings-ssv: "alpha" "beta" 'gamma' 'delta';

.csv {
  output: $quoted-strings-csv;
  output: #{$quoted-strings-csv};
  output: "[#{$quoted-strings-csv}]";
  output: "#{$quoted-strings-csv}";
  output: "["#{$quoted-strings-csv}"]";
  output: '#{$quoted-strings-csv}';
  output: "['#{$quoted-strings-csv}']";
}

.ssv {
  output: $quoted-strings-ssv;
  output: #{$quoted-strings-ssv};
  output: "[#{$quoted-strings-ssv}]";
  output: "#{$quoted-strings-ssv}";
  output: "["#{$quoted-strings-ssv}"]";
  output: '#{$quoted-strings-ssv}';
  output: "['#{$quoted-strings-ssv}']";
}
```
ruby-sass 3.4.11 output, according to SassMeister
```css
.csv {
  output: "alpha", "beta", "gamma", "delta";
  output: alpha, beta, gamma, delta;
  output: "[alpha, beta, gamma, delta]";
  output: "alpha, beta, gamma, delta";
  output: "["alpha, beta, gamma, delta"]";
  output: "alpha, beta, gamma, delta";
  output: "['alpha, beta, gamma, delta']"; }

.ssv {
  output: "alpha" "beta" "gamma" "delta";
  output: alpha beta gamma delta;
  output: "[alpha beta gamma delta]";
  output: "alpha beta gamma delta";
  output: "["alpha beta gamma delta"]";
  output: "alpha beta gamma delta";
  output: "['alpha beta gamma delta']"; }
```
output from sassc: 3.1.0-2-g2033 / libsass: 3.1.0-76-g71448:
```css
.csv {
  output: "alpha", "beta", 'gamma', 'delta';
  output: "alpha", "beta", 'gamma', 'delta';
  output: "["alpha", "beta", 'gamma', 'delta']";
  output: ""alpha", "beta", 'gamma', 'delta'";
  output: "[""alpha", "beta", 'gamma', 'delta'"]";
  output: '"alpha", "beta", 'gamma', 'delta'';
  output: "['"alpha", "beta", 'gamma', 'delta'']"; }

.ssv {
  output: "alpha" "beta" 'gamma' 'delta';
  output: "alpha" "beta" 'gamma' 'delta';
  output: "["alpha" "beta" 'gamma' 'delta']";
  output: ""alpha" "beta" 'gamma' 'delta'";
  output: "[""alpha" "beta" 'gamma' 'delta'"]";
  output: '"alpha" "beta" 'gamma' 'delta'';
  output: "['"alpha" "beta" 'gamma' 'delta'']"; }
```

**Comment by mgreter** [comment]
This may already be solved by my current WIP branch: https://github.com/sass/libsass/pull/839
Will have time in a few hours to run your test against it.

**Comment by xzyfer** [comment]
Specs added https://github.com/sass/sass-spec/pull/252

**Comment by mgreter** [comment]
OK, will have a little time tomorrow to debug this issue a little more. But I already have some good news. As I said, I thought I have adressed this issue already, but as it turns out, only if the lists are declared inside the interpolations directly and not via variables. So here's a little peak:

Input:
```CSS
.csv {
  output: "alpha", "beta", 'gamma', 'delta';
  output: #{"alpha", "beta", 'gamma', 'delta'};
  output: "[#{"alpha", "beta", 'gamma', 'delta'}]";
  output: "#{"alpha", "beta", 'gamma', 'delta'}";
  output: '#{"alpha", "beta", 'gamma', 'delta'}';
  output: "['#{"alpha", "beta", 'gamma', 'delta'}']";
}

.ssv {
  output: "alpha" "beta" 'gamma' 'delta';
  output: #{"alpha" "beta" 'gamma' 'delta'};
  output: "[#{"alpha" "beta" 'gamma' 'delta'}]";
  output: "#{"alpha" "beta" 'gamma' 'delta'}";
  output: '#{"alpha" "beta" 'gamma' 'delta'}';
  output: "['#{"alpha" "beta" 'gamma' 'delta'}']";
}
```

Results:
```
# +---+--------------------------------------------------+--------------------------------------------------+
# | Ln|Got                                               |Expected                                          |
# +---+--------------------------------------------------+--------------------------------------------------+
# |  1|'.csv {                                           |'.csv {                                           |
# |  2|  output: "alpha", "beta", \'gamma\', \'delta\';  |  output: "alpha", "beta", \'gamma\', \'delta\';  |
# |  3|  output: alpha, beta, gamma, delta;              |  output: alpha, beta, gamma, delta;              |
# |  4|  output: "[alpha, beta, gamma, delta]";          |  output: "[alpha, beta, gamma, delta]";          |
# |  5|  output: "alpha, beta, gamma, delta";            |  output: "alpha, beta, gamma, delta";            |
# *  6|  output: \'alpha, beta, gamma, delta\';          |  output: "alpha, beta, gamma, delta";            *
# |  7|  output: "[\'alpha, beta, gamma, delta\']"; }    |  output: "[\'alpha, beta, gamma, delta\']"; }    |
# |  8|.ssv {                                            |.ssv {                                            |
# |  9|  output: "alpha" "beta" \'gamma\' \'delta\';     |  output: "alpha" "beta" \'gamma\' \'delta\';     |
# | 10|  output: alpha beta gamma delta;                 |  output: alpha beta gamma delta;                 |
# | 11|  output: "[alpha beta gamma delta]";             |  output: "[alpha beta gamma delta]";             |
# | 12|  output: "alpha beta gamma delta";               |  output: "alpha beta gamma delta";               |
# * 13|  output: \'alpha beta gamma delta\';             |  output: "alpha beta gamma delta";               *
# | 14|  output: "[\'alpha beta gamma delta\']"; }       |  output: "[\'alpha beta gamma delta\']"; }       |
# | 15|'                                                 |'                                                 |
# +---+--------------------------------------------------+--------------------------------------------------+
```

So this already seems to do the right thing, but the output with your actual test (with the lists comming from variables) is still off by quite a lot:

```
# +---+------------------------------------------------------------------------+------------------------------------------------+
# | Ln|Got                                                                     |Expected                                        |
# +---+------------------------------------------------------------------------+------------------------------------------------+
# |  1|'.csv {                                                                 |'.csv {                                         |
# |  2|  output: "alpha", "beta", "gamma", "delta";                            |  output: "alpha", "beta", "gamma", "delta";    |
# *  3|  output: "alpha", "beta", "gamma", "delta";                            |  output: alpha, beta, gamma, delta;            *
# *  4|  output: \'["alpha", "beta", "gamma", "delta"]\';                      |  output: "[alpha, beta, gamma, delta]";        *
# *  5|  output: \'"alpha", "beta", "gamma", "delta"\';                        |  output: "alpha, beta, gamma, delta";          *
# *  6|  output: \'"alpha", "beta", "gamma", "delta"\';                        |  output: "alpha, beta, gamma, delta";          *
# *  7|  output: "[\'\\"alpha\\", \\"beta\\", \\"gamma\\", \\"delta\\"\']"; }  |  output: "[\'alpha, beta, gamma, delta\']"; }  *
# |  8|.ssv {                                                                  |.ssv {                                          |
# |  9|  output: "alpha" "beta" "gamma" "delta";                               |  output: "alpha" "beta" "gamma" "delta";       |
# * 10|  output: "alpha" "beta" "gamma" "delta";                               |  output: alpha beta gamma delta;               *
# * 11|  output: \'["alpha" "beta" "gamma" "delta"]\';                         |  output: "[alpha beta gamma delta]";           *
# * 12|  output: \'"alpha" "beta" "gamma" "delta"\';                           |  output: "alpha beta gamma delta";             *
# * 13|  output: \'"alpha" "beta" "gamma" "delta"\';                           |  output: "alpha beta gamma delta";             *
# * 14|  output: "[\'\\"alpha\\" \\"beta\\" \\"gamma\\" \\"delta\\"\']"; }     |  output: "[\'alpha beta gamma delta\']"; }     *
# | 15|'                                                                       |'                                               |
# +---+------------------------------------------------------------------------+------------------------------------------------+
```

But I hope this test actually gets me a little bit closer to the really correct solution. But IMHO the first output I showed is already quite a big improvement :smile: But be aware that I really cannot give any ETA on this big refactoring, since it involes the parser, the evaluation and the output part of the code!

@xzyfer I will probably have quite a lot of tests to add, once I get this all working ...

**Comment by mgreter** [comment]
It looks like sass "forgets" the original quotes when interpolating lists. Once I disabled that in my code, your example actually passes perfectly with my local branch :smile: Now I need to test how this behaves with nested interpolates inside or outside quotes and other variations to be sure about the rules ...

**Comment by mgreter** [comment]
Added a few auto generated tests to cover this issue a bit more in-depth.
https://github.com/mgreter/sass-specs/tree/master/suite/parser/interpolate
And with a few tweaks in my current WIP branch I was able to pass them all!
Just don't get too excited! Still a lot of work to clean up the mess I made in that branch!

**Comment by lunelson** [comment]
This is great news @mgreter, thanks. It does seem to be as you say, that ruby sass interpolates a list by unquoting everything to make it one string, and if this in turn is quoted , the outermost will be normalized to double quotes. It might be worth asking @nex3 or @chriseppstein to chime in though, as I seem to remember my test gave different output in sass 3.3.x vs 3.4.x , so I think they updated this logic recently.

**Comment by mgreter** [comment]
I'm currently (re-)generating the expected results with ruby sass 3.4.9. Another thing I have not solved 100% correctly yet is handling of escape sequences and it seems as soon as I fix one side it breaks on the other (or interpolation fails). So there is still a lot to do until I want to get this all merged. By the way, I also found some cases where ruby sass chokes on the input, while my WIP branch happily processes it:

```
.result {
  dquoted: "#{#{literal}}";
}
```

**Comment by nex3** [comment]
@lunelson I'm not sure I understand the behavior you're discussing.

@mgreter can you file an issue for that?

**Comment by lunelson** [comment]
@nex3 I was inferring that since—according to Sassmeister—the code in the first post renders differently in sass 3.3 vs 3.4, something might have explicitly changed there; but in any case that you could probably clarify in plain terms what the logic is behind sass' interpolation of—for example—lists of quoted items

**Comment by nex3** [comment]
When you write `#{$foo}` anywhere, the normal conversion of `$foo` to a CSS string takes place with the caveat that if `$foo` or anything recursively contained in `$foo` is a string, its quotes are removed and any newlines are folded into spaces.

**Comment by lunelson** [comment]
Thanks for that @nex3

**Comment by mgreter** [comment]
This should be solved by #910!
